rust/operator-binary/src/restart_controller/pod.rs

```diff
@@ -0,0 +1,127 @@
+use std::{sync::Arc, time::Duration};
+
+use futures::StreamExt;
+use snafu::{OptionExt, ResultExt, Snafu};
+use stackable_operator::{
+    client::Client,
+    k8s_openapi::{
+        api::core::v1::Pod,
+        chrono::{self, DateTime, FixedOffset, Utc},
+    },
+    kube::{
+        self,
+        api::{EvictParams, ListParams},
+        core::DynamicObject,
+        runtime::{
+            controller::{Action, Context},
+            reflector::ObjectRef,
+            Controller,
+        },
+    },
+    logging::controller::{report_controller_reconciled, ReconcilerError},
+};
+use strum::{EnumDiscriminants, IntoStaticStr};
+
+struct Ctx {
+    client: Client,
+}
+
+#[derive(Snafu, Debug, EnumDiscriminants)]
+#[strum_discriminants(derive(IntoStaticStr))]
+enum Error {
+    #[snafu(display("Pod has no name"))]
+    PodHasNoName,
+    #[snafu(display(
+        "failed to parse expiry timestamp annotation ({annotation:?}: {value:?}) as RFC 3999"
+    ))]
+    UnparseableExpiryTimestamp {
+        source: chrono::ParseError,
+        annotation: String,
+        value: String,
+    },
+    #[snafu(display("failed to evict Pod"))]
+    EvictPod { source: kube::Error },
+}
+
+impl ReconcilerError for Error {
+    fn category(&self) -> &'static str {
+        ErrorDiscriminants::from(self).into()
+    }
+
+    fn secondary_object(&self) -> Option<ObjectRef<DynamicObject>> {
+        match self {
+            Error::PodHasNoName => None,
+            Error::UnparseableExpiryTimestamp {
+                source: _,
+                annotation: _,
+                value: _,
+            } => None,
+            Error::EvictPod { source: _ } => None,
```

Comment:
Nit: It would be great if we could include the `ObjectRef<Pod>` inside the two errors and return them here
Also we could simply the first matches to `Error::UnparseableExpiryTimestamp { .. }` and `Error::EvictPod { .. }`

Reply:
The `ObjectRef` to the main object is already added by `kube::runtime` when it wraps the error.